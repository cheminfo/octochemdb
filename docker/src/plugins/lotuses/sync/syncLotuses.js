import md5 from 'md5';

import getLastDocumentImported from '../../../sync/http/utils/getLastDocumentImported.js';
import getLastFileSync from '../../../sync/http/utils/getLastFileSync.js';
import debugLibrary from '../../../utils/Debug.js';
import { getTaxonomiesForLotuses } from '../../activesOrNaturals/utils/utilsTaxonomies/getTaxonomiesForLotuses.js';
import { taxonomySynonyms } from '../../activesOrNaturals/utils/utilsTaxonomies/taxonomySynonyms.js';

import { parseLotuses } from './utils/parseLotuses.js';
/**
 * @description syncLotuses - synchronize lotuses collection from lotus database
 * @param {*} connection - mongo connection
 * @returns {Promise} returns collection lotuses
 */
export async function sync(connection) {
  const debug = debugLibrary('syncLotuses');
  let options = {
    collectionSource: process.env.LOTUS_SOURCE,
    destinationLocal: `${process.env.ORIGINAL_DATA_PATH}/lotuses/full`,
    collectionName: 'lotuses',
    filenameNew: 'lotuses',
    extensionNew: 'zip',
  };
  try {
    let sources;
    let lastFile;
    if (process.env.NODE_ENV === 'test') {
      lastFile = `${process.env.LOTUS_SOURCE_TEST}`;
      sources = [lastFile];
    } else {
      // get last file from lotus database
      lastFile = await getLastFileSync(options);
      // get sources, progress and lotuses collection
      sources = [lastFile.replace(`${process.env.ORIGINAL_DATA_PATH}`, '')];
    }
    debug(lastFile);
    // get sources, progress and lotuses collection
    const progress = await connection.getProgress('lotuses');
    let isTimeToUpdate = false;
    if (
      progress.dateEnd !== 0 &&
      Date.now() - progress.dateEnd >
        Number(process.env.LOTUS_UPDATE_INTERVAL) * 24 * 60 * 60 * 1000 &&
      md5(JSON.stringify(sources)) !== progress.sources
    ) {
      progress.dateStart = Date.now();
      await connection.setProgress(progress);
      isTimeToUpdate = true;
    }
    // get last document imported
    const lastDocumentImported = await getLastDocumentImported(
      connection,
      progress,
      options.collectionName,
    );

    // define counter
    let counter = 0;
    let imported = 0;
    let start = Date.now();
    if (
      lastDocumentImported === null ||
      ((md5(JSON.stringify(sources)) !== progress.sources ||
        progress.state !== 'updated') &&
        isTimeToUpdate)
    ) {
      const collection = await connection.getCollection('lotuses');
      // get old to new taxonomies ids mapping
      const oldToNewTaxIDs = await taxonomySynonyms();
      // get taxonomies collection
      const collectionTaxonomies = await connection.getCollection('taxonomies');
      // get logs
      const logs = await connection.getImportationLog({
        collectionName: options.collectionName,
        sources,
        startSequenceID: progress.seq,
      });
      // define file inside zip folder to use for importation
      let fileName = 'lotusUniqueNaturalProduct.bson';
      // create temporary collection
      const temporaryCollection = await connection.getCollection(
        `${options.collectionName}_tmp`,
      );
      debug(`Start parsing: ${fileName}`);
      // set progress state to updating
      progress.state = 'updating';
      await connection.setProgress(progress);
      // parse lotuses
      for await (const entry of parseLotuses(lastFile, fileName, connection)) {
        counter++;
        if (process.env.TEST === 'true' && counter > 20) break;

        if (Date.now() - start > Number(process.env.DEBUG_THROTTLING)) {
          debug(`Processing: counter: ${counter} - imported: ${imported}`);
          start = Date.now();
        }
        /// Normalize Taxonomies
        if (entry.data.taxonomies) {
          let taxonomies = await getTaxonomiesForLotuses(
            entry,
            collectionTaxonomies,
            oldToNewTaxIDs,
          );
          entry.data.taxonomies = taxonomies;
        }
        entry._seq = ++progress.seq;
        await temporaryCollection.updateOne(
          { _id: entry._id },
          { $set: entry },
          { upsert: true },
        );
        imported++;
      }
      // Temporary collection replace old collection
      await temporaryCollection.rename(options.collectionName, {
        dropTarget: true,
      });

      // update Logs in importationLogs collection
      logs.dateEnd = Date.now();
      logs.endSequenceID = progress.seq;
      logs.status = 'updated';
      await connection.updateImportationLog(logs);

      //Update progress in admin collection
      progress.sources = md5(JSON.stringify(sources));
      progress.dateEnd = Date.now();
      progress.state = 'updated';
      await connection.setProgress(progress);
      // Indexing of collection properties
      await collection.createIndex({ 'data.ocl.noStereoTautomerID': 1 });
      await collection.createIndex({ _seq: 1 });

      debug(`${imported} compounds processed`);
    } else {
      debug(`file already processed`);
    }
  } catch (e) {
    if (connection) {
      debug(e.message, { collection: 'lotuses', connection, stack: e.stack });
    }
  }
}
