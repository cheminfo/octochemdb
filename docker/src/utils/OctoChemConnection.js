import delay from 'delay';
import dotenv from 'dotenv';
import md5 from 'md5';
import { MongoClient } from 'mongodb';

import debugLibrary from './Debug.js';

dotenv.config();
const debug = debugLibrary('OctoChemConnection');

export function OctoChemConnection() {
  this.client = new MongoClient(process.env.MONGODB_URL, {
    keepAlive: true,
    connectTimeoutMS: 6 * 60 * 60 * 1000,
    socketTimeoutMS: 6 * 60 * 60 * 1000,
  });
}

OctoChemConnection.prototype.close = async function close() {
  if (this.connection) return this.connection.close();
  return undefined;
};

OctoChemConnection.prototype.getCollectionNames =
  async function getCollectionNames() {
    return (await (await this.getDatabase()).listCollections().toArray()).map(
      (entry) => entry.name,
    );
  };

OctoChemConnection.prototype.getCollection = async function getCollection(
  collectionName,
) {
  return (await this.getDatabase()).collection(collectionName);
};

OctoChemConnection.prototype.getImportationLog =
  async function getImportationLog(options) {
    const { collectionName, sources, startSequenceID } = options;
    const logsCollection = await this.getImportationLogsCollection();
    const sourcesHash = md5(JSON.stringify(sources));
    const _id = sourcesHash;
    let logs = await logsCollection.find({ _id }).next();
    if (logs === null) {
      logs = {
        _id,
        collectionName,
        sources,
        sourcesHash,
        dateStart: Date.now(),
        dateEnd: Date.now(),
        startSequenceID,
        endSequenceID: 0,
        status: 'updating',
      };
      await logsCollection.insertOne(logs);
    }
    return logs;
  };
OctoChemConnection.prototype.updateImportationLog =
  async function updateImportationLog(logs) {
    const collection = await this.getImportationLogsCollection();
    logs.dateEnd = Date.now();
    await collection.replaceOne({ _id: logs._id }, logs);
  };

OctoChemConnection.prototype.getImportationLogsCollection =
  async function getCollection() {
    return (await this.getDatabase()).collection('importationLogs');
  };

OctoChemConnection.prototype.getAdminCollection =
  async function getAdminCollection() {
    return this.getCollection('admin');
  };

OctoChemConnection.prototype.setProgress = async function setProgress(
  progress,
) {
  const collection = await this.getAdminCollection();
  await collection.replaceOne({ _id: progress._id }, progress);
};

OctoChemConnection.prototype.getProgress = async function getProgress(
  collectionName,
) {
  const adminCollection = await this.getAdminCollection();
  const _id = `${collectionName}_progress`;
  let progress = await adminCollection.find({ _id }).next();
  if (progress === null) {
    progress = {
      _id,
      state: 'updating',
      seq: 0,
      dateStart: Date.now(),
      dateEnd: 0,
    };
    await adminCollection.insertOne(progress);
  }
  return progress;
};

OctoChemConnection.prototype.getDatabase = async function getDatabase() {
  while (true) {
    try {
      await this.init();
      break;
    } catch (e) {
      debug('Connection to mongo failed, waiting 5s');
      debug(e.stack);
      await delay(5000);
    }
  }
  return this.connection.db(process.env.MONGO_DB_NAME);
};

OctoChemConnection.prototype.init = async function init() {
  if (this.connection) return;

  debug(`Trying to connect to: ${process.env.MONGODB_URL}`);

  this.connection = await this.client.connect();
  debug('Got DB connection');
};
