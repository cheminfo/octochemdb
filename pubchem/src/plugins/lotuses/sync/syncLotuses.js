import md5 from 'md5';

import getLastDocumentImported from '../../../sync/http/utils/getLastDocumentImported.js';
import getLastFileSync from '../../../sync/http/utils/getLastFileSync.js';
import Debug from '../../../utils/Debug.js';
import { getTaxonomiesForLotuses } from '../../activesOrNaturals/utils/utilsTaxonomies/getTaxonomiesForLotuses.js';
import { taxonomySynonyms } from '../../activesOrNaturals/utils/utilsTaxonomies/taxonomySynonyms.js';

import { parseLotuses } from './utils/parseLotuses.js';
/**
 * @description syncLotuses - synchronize lotuses collection from lotus database
 * @param {*} connection - mongo connection
 * @returns {Promise} returns collection lotuses
 */
export async function sync(connection) {
  const debug = Debug('syncLotuses');
  let options = {
    collectionSource: process.env.LOTUS_SOURCE,
    destinationLocal: `${process.env.ORIGINAL_DATA_PATH}/lotuses/full`,
    collectionName: 'lotuses',
    filenameNew: 'lotuses',
    extensionNew: 'zip',
  };
  try {
    // get last file from lotus database
    const lastFile = await getLastFileSync(options);
    // get sources, progress and lotuses collection
    const sources = [lastFile.replace(process.env.ORIGINAL_DATA_PATH, '')];
    const progress = await connection.getProgress('lotuses');
    const collection = await connection.getCollection('lotuses');
    // get old to new taxonomies ids mapping
    const oldToNewTaxIDs = await taxonomySynonyms();
    // get taxonomies collection
    const collectionTaxonomies = await connection.getCollection('taxonomies');
    // get logs
    const logs = await connection.getImportationLog({
      collectionName: options.collectionName,
      sources,
      startSequenceID: progress.seq,
    });
    // get last document imported
    const lastDocumentImported = await getLastDocumentImported(
      connection,
      progress,
      options.collectionName,
    );
    // define file inside zip folder to use for importation
    let fileName = 'lotusUniqueNaturalProduct.bson';
    // define counter
    let counter = 0;
    let imported = 0;
    let start = Date.now();
    if (
      lastDocumentImported === null ||
      md5(JSON.stringify(sources)) !== progress.sources ||
      progress.state !== 'updated'
    ) {
      // create temporary collection
      const temporaryCollection = await connection.getCollection(
        'temporaryLotuses',
      );
      debug(`Start parsing: ${fileName}`);
      // set progress state to updating
      progress.state = 'updating';
      await connection.setProgress(progress);
      // parse lotuses
      for await (const entry of parseLotuses(lastFile, fileName, connection)) {
        counter++;
        if (process.env.TEST === 'true' && counter > 20) break;

        if (
          Date.now() - start >
          Number(process.env.DEBUG_THROTTLING || 10000)
        ) {
          debug(`Processing: counter: ${counter} - imported: ${imported}`);
          start = Date.now();
        }
        /// Normalize Taxonomies
        if (entry.data.taxonomies) {
          let taxonomies = await getTaxonomiesForLotuses(
            entry,
            collectionTaxonomies,
            oldToNewTaxIDs,
          );
          entry.data.taxonomies = taxonomies;
        }
        entry._seq = ++progress.seq;
        await temporaryCollection.updateOne(
          { _id: entry._id },
          { $set: entry },
          { upsert: true },
        );
        imported++;
      }
      // Temporary collection replace old collection
      await temporaryCollection.rename(options.collectionName, {
        dropTarget: true,
      });

      // update Logs in importationLogs collection
      logs.dateEnd = Date.now();
      logs.endSequenceID = progress.seq;
      logs.status = 'updated';
      await connection.updateImportationLog(logs);

      //Update progress in admin collection
      progress.sources = md5(JSON.stringify(sources));
      progress.dateEnd = Date.now();
      progress.state = 'updated';
      await connection.setProgress(progress);
      // Indexing of collection properties
      await collection.createIndex({ _id: 1 });
      await collection.createIndex({ _seq: 1 });
      await collection.createIndex({ 'data.ocl.noStereoID': 1 });
      debug(`${imported} compounds processed`);
    } else {
      debug(`file already processed`);
    }
  } catch (e) {
    if (connection) {
      debug(e.message, { collection: 'lotuses', connection, stack: e.stack });
    }
  }
}
