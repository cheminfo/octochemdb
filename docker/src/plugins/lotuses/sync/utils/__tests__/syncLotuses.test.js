import delay from 'delay';
import { test, expect } from 'vitest';

import { OctoChemConnection } from '../../../../../utils/OctoChemConnection.js';
import { sync } from '../../syncLotuses';

test(
  'syncLotuses',
  async () => {
    const connection = new OctoChemConnection();
    let colllectionList = await connection.getCollectionNames();
    while (!colllectionList.includes('taxonomies')) {
      await delay(1000);
      colllectionList = await connection.getCollectionNames();
    }
    await sync(connection);
    const collection = await connection.getCollection('lotuses');
    const collectionEntry = await collection
      .find({ _id: 'LTS0257199' })
      .limit(1);
    const result = await collectionEntry.next();
    if (result?._seq) {
      delete result._seq;
    }
    expect(result).toMatchSnapshot();
    await connection.close();
  },
  { timeout: 30000 },
);
