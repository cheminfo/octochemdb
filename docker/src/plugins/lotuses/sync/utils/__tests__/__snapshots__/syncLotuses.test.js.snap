// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`syncLotuses 1`] = `
{
  "_id": "LTS0257199",
  "data": {
    "iupacName": "(2R,3S,4R,5R,6R)-2-(hydroxymethyl)-6-{[(1R,4S,5S,8R,9R,12S,13S,16S)-8-[(2R,4S)-4-methoxy-6-methylhept-5-en-2-yl]-5,9,17,17-tetramethyl-18-oxapentacyclo[10.5.2.0¹,¹³.0⁴,¹².0⁵,⁹]nonadec-2-en-16-yl]oxy}oxane-3,4,5-triol",
    "ocl": {
      "idCode": "ekTpA@@@LAEMGLn\\dTTRbRfLbteRrRTfbqbtRthdRjZFFfNnAQjjjjjjjfjjjjjijjkrBHPaTKaRSrUPZSKZWHEXqe@@",
      "noStereoTautomerID": "ekTpA@@@LAEMGLn\\dTTRbRfLbteRrRTfbqbtRthdRjZFFfNnAQjjjjjjjfjjjjjijjh@@",
    },
    "taxonomies": [
      {
        "class": "Magnoliopsida",
        "dbRef": {
          "$id": "LTS0257199",
          "$ref": "lotuses",
        },
        "family": "Cucurbitaceae",
        "genus": "Momordica",
        "kingdom": "Viridiplantae",
        "phylum": "Streptophyta",
        "species": "Momordica charantia",
      },
    ],
  },
}
`;
